### PR TITLE
feat: TLS implementation in gRPC sink

### PR DESCRIPTION
**Pull request by Vaishnavi190900**
Implemented TLS support in gRPC sink.
also added 2 configurations for gRPC sink:-
1. SINK_GRPC_TLS_ENABLE → indicates whether the sink needs to be connected over TLS. Default value is false
2. SINK_GRPC_ROOT_CA → The ca-certificates of GOTO domain *.gojek.gcp

### CODE REVIEW

**Review by manigandand — APPROVED** [review]

File: src/main/java/com/gotocompany/firehose/sink/grpc/GrpcSinkFactory.java [hunk 1]
```diff
@@ -21,23 +31,64 @@
...
 
+    private static final String CERT_DIRECTORY = "/usr/local/share/ca-certificates";
+
+    private static File createCertFileFromBase64(String base64Cert) throws IOException {
+        byte[] decodedBytes = Base64.getDecoder().decode(base64Cert);
+        File certDirectory = new File(CERT_DIRECTORY);
+        if (!certDirectory.exists()) {
+            certDirectory.mkdirs();
+        }
+        File certFile = Files.createTempFile(certDirectory.toPath(), "goto-root-ca", ".crt").toFile();
```
Comment (sbchaos) [inline src/main/java/com/gotocompany/firehose/sink/grpc/GrpcSinkFactory.java:45]: Can we have a generic name for the cert file if it does not break things.
`goto-root-ca` -> `root-ca` or `client-cert`

File: src/main/java/com/gotocompany/firehose/sink/grpc/GrpcSinkFactory.java [hunk 2]
```diff
@@ -21,23 +31,64 @@
...
+        log.info(
+                "Created temporary file at {} to {}", certFile.getAbsolutePath(), certFile.toPath()
```
Comment (sbchaos) [inline src/main/java/com/gotocompany/firehose/sink/grpc/GrpcSinkFactory.java:47]: If we are logging the absolute path, do we need to get the path again ?
Does absolute path contain the file name.

File: src/main/java/com/gotocompany/firehose/sink/grpc/GrpcSinkFactory.java [hunk 3]
```diff
@@ -21,23 +31,64 @@
...
+            fos.write(decodedBytes);
+        }
+        return certFile;
+    }
+
+    private static SslContext buildClientSslContext(String base64Cert) {
+        try {
+            File certFile = createCertFileFromBase64(base64Cert);
+            return GrpcSslContexts
+                    .configure(SslContextBuilder.forClient().trustManager(certFile))
```
Comment (sbchaos) [inline src/main/java/com/gotocompany/firehose/sink/grpc/GrpcSinkFactory.java:59]: Can we please simplify this ?
Multiple function calls inside a builder.

File: src/main/java/com/gotocompany/firehose/sink/grpc/GrpcSinkFactory.java [hunk 4]
```diff
@@ -21,23 +31,64 @@
...
-
-        ManagedChannel managedChannel = ManagedChannelBuilder.forAddress(grpcConfig.getSinkGrpcServiceHost(), grpcConfig.getSinkGrpcServicePort())
-                        .keepAliveTime(grpcConfig.getSinkGrpcArgKeepaliveTimeMS(), TimeUnit.MILLISECONDS)
-                        .keepAliveTimeout(grpcConfig.getSinkGrpcArgKeepaliveTimeoutMS(), TimeUnit.MILLISECONDS)
-                        .usePlaintext().build();
-
-        GrpcClient grpcClient = new GrpcClient(new FirehoseInstrumentation(statsDReporter, GrpcClient.class), grpcConfig, managedChannel, stencilClient);
-        firehoseInstrumentation.logInfo("GRPC connection established");
-
+        boolean isTlsEnable = grpcConfig.getSinkGrpcTlsEnable();
```
Comment (sbchaos) [inline src/main/java/com/gotocompany/firehose/sink/grpc/GrpcSinkFactory.java:72]: nit: `isTlsEnable` -> `isTlsEnabled`

File: src/main/java/com/gotocompany/firehose/sink/grpc/GrpcSinkFactory.java [hunk 5]
```diff
@@ -21,23 +31,64 @@
...
+                    .keepAliveTimeout(grpcConfig.getSinkGrpcArgKeepaliveTimeoutMS(), TimeUnit.MILLISECONDS)
+                    .sslContext(sslContext)
+                    .build();
+            grpcClient = new GrpcClient(new FirehoseInstrumentation(statsDReporter, GrpcClient.class), grpcConfig, managedChannel, stencilClient);
+        } else {
+            ManagedChannel managedChannel = ManagedChannelBuilder.forAddress(grpcConfig.getSinkGrpcServiceHost(), grpcConfig.getSinkGrpcServicePort())
+                    .keepAliveTime(grpcConfig.getSinkGrpcArgKeepaliveTimeMS(), TimeUnit.MILLISECONDS)
+                    .keepAliveTimeout(grpcConfig.getSinkGrpcArgKeepaliveTimeoutMS(), TimeUnit.MILLISECONDS)
+                    .usePlaintext()
+                    .build();
```
Comment (sbchaos) [inline src/main/java/com/gotocompany/firehose/sink/grpc/GrpcSinkFactory.java:89]: A lot of configuration seems to be common, can we extract out the common parts out of the if-else block.
eg 
```
   ManagedChannelBuilder managedChannelBuilder = NettyChannelBuilder
                    .forAddress(grpcConfig.getSinkGrpcServiceHost(), grpcConfig.getSinkGrpcServicePort())
                    .keepAliveTime(grpcConfig.getSinkGrpcArgKeepaliveTimeMS(), TimeUnit.MILLISECONDS)
                    .keepAliveTimeout(grpcConfig.getSinkGrpcArgKeepaliveTimeoutMS(), TimeUnit.MILLISECONDS);

      if (isTlsEnabled) {
          String base64Cert = grpcConfig.getSinkGrpcRootCA();
          SslContext sslContext = buildClientSslContext(base64Cert);
          firehoseInstrumentation.logInfo("SSL Context created successfully.");
          ManagedChannelBuilder.sslContext(sslContext);
      } else {
          ManagedChannelBuilder.usePlaintext();
      }
    
   grpcClient = new GrpcClient(new FirehoseInstrumentation(statsDReporter, GrpcClient.class), grpcConfig, managedChannelBuilder.build(), stencilClient);
```

Comment (Vaishnavi190900) [reply src/main/java/com/gotocompany/firehose/sink/grpc/GrpcSinkFactory.java:89]: @sbchaos Initially, I had structured the code similarly by creating a common ManagedChannelBuilder instance, but encountered an error when attempting to reuse it across the if-else branches. Consequently, I opted to use separate variables within each block.

File: src/main/java/com/gotocompany/firehose/sink/grpc/GrpcSinkFactory.java [hunk 6]
```diff
@@ -21,24 +29,56 @@
  * using the configurations supplied and invoke {@see #create(Map < String, String > configuration, StatsDClient client)}
  * to obtain the GrpcSink sink implementation.
  */
+@Slf4j
 public class GrpcSinkFactory {
 
+    private static final String CERT_DIRECTORY = "/usr/local/share/ca-certificates";
```
Comment (rajuGT) [inline src/main/java/com/gotocompany/firehose/sink/grpc/GrpcSinkFactory.java:35]: name it to PERSIST_CERT_DIRECTORY_PATH

File: src/main/java/com/gotocompany/firehose/sink/grpc/GrpcSinkFactory.java [hunk 7]
```diff
@@ -21,24 +29,56 @@
...
+
+    private static SslContext buildClientSslContext(String base64Cert) {
+        try {
+            File certFile = createCertFileFromBase64(base64Cert);
+            SslContextBuilder sslContextBuilder = SslContextBuilder.forClient().trustManager(certFile);
+            return GrpcSslContexts
+                    .configure(sslContextBuilder)
+                    .build();
+        } catch (IOException e) {
+            throw new RuntimeException(e);
```
Comment (rajuGT) [inline src/main/java/com/gotocompany/firehose/sink/grpc/GrpcSinkFactory.java:59]: throw new RuntimeException(e);, RuntimeException("failed to create ssl ......", e)

**Comment by rajuGT** [comment]
In the same MR, update firehose doc for this feature.

**Review by rajuGT — CHANGES_REQUESTED** [review]
minor feedbacks

File: src/main/java/com/gotocompany/firehose/sink/grpc/GrpcSinkFactory.java [hunk 8]
```diff
@@ -21,24 +29,57 @@
...
+
+    private static File createCertFileFromBase64(String base64Cert) throws IOException {
+        byte[] decodedBytes = Base64.getDecoder().decode(base64Cert);
+        File certDirectory = new File(PERSIST_CERT_DIRECTORY_PATH);
+        if (!certDirectory.exists()) {
+            certDirectory.mkdirs();
+        }
+        File certFile = Files.createTempFile(certDirectory.toPath(), "client-cert", ".crt").toFile();
+        log.info("Created certificate file at {}", certFile.getAbsolutePath());
+        try (FileOutputStream fos = new FileOutputStream(certFile)) {
```
Comment (rajuGT) [inline src/main/java/com/gotocompany/firehose/sink/grpc/GrpcSinkFactory.java:45]: As we are not handling exception within the block, check if this `try` is required.

**Review by rajuGT — CHANGES_REQUESTED** [review]

**Review by rajuGT — APPROVED** [review]

File: src/main/java/com/gotocompany/firehose/sink/grpc/client/GrpcClient.java [hunk 9]
```diff
@@ -58,18 +59,20 @@ public DynamicMessage execute(byte[] logMessage, Headers headers) {
...
         } catch (StatusRuntimeException sre) {
-            firehoseInstrumentation.logError("gRPC call failed with error message: {}", sre.getMessage());
+            if (sre.getStatus().getCode() == Status.Code.UNAVAILABLE) {
+                firehoseInstrumentation.logError("gRPC configurations are incorrect: {}", sre.getMessage());
+            } else {
+                firehoseInstrumentation.logError("gRPC call failed with error message: {}", sre.getMessage());
+            }
             firehoseInstrumentation.incrementCounter(Metrics.SINK_GRPC_ERROR_TOTAL,  "status=" + sre.getStatus().getCode());
         } catch (Exception e) {
             firehoseInstrumentation.logError("gRPC call failed with error message: {}", e.getMessage());
```
Comment (rajuGT) [inline src/main/java/com/gotocompany/firehose/sink/grpc/client/GrpcClient.java:78]: do we need to include the test resource binary file which is newly added in the MR? If not, we can update .gitignore file.

**Review by rajuGT — APPROVED** [review]

File: src/main/java/com/gotocompany/firehose/sink/grpc/GrpcSinkFactory.java [hunk 10]
```diff
@@ -26,22 +38,64 @@
...
+            fos.write(decodedBytes);
+        }
+        return certFile;
+    }
+
+    private static SslContext buildClientSslContext(String base64Cert) {
+        try {
+            File certFile = createCertFileFromBase64(base64Cert);
```
Comment (ekawinataa) [inline src/main/java/com/gotocompany/firehose/sink/grpc/GrpcSinkFactory.java:62]: Just wondering whether can we skip the temporary file and directly pass the cert as `ByteArrayInputStream` instead. Maybe something like this 

`    private static SslContext buildClientSslContext(String base64Cert) {
        try {
            byte[] decodedBytes = Base64.getDecoder().decode(base64Cert);
            ByteArrayInputStream certInputStream = new ByteArrayInputStream(decodedBytes);
            SslContextBuilder sslContextBuilder = SslContextBuilder.forClient().trustManager(certInputStream);
            return GrpcSslContexts
                    .configure(sslContextBuilder)
                    .build();
        } catch (IOException e) {
            throw new RuntimeException("Failed to build SSL context due to an IO error while creating the certificate file.", e);
        }
    }`

File: src/main/java/com/gotocompany/firehose/sink/grpc/GrpcSinkFactory.java [hunk 11]
```diff
@@ -26,22 +38,64 @@
...
+        }
+        File certFile = Files.createTempFile(certDirectory.toPath(), "client-cert", ".crt").toFile();
+        log.info("Created certificate file at {}", certFile.getAbsolutePath());
+        try (FileOutputStream fos = new FileOutputStream(certFile)) {
+            fos.write(decodedBytes);
+        }
+        return certFile;
+    }
+
+    private static SslContext buildClientSslContext(String base64Cert) {
```
Comment (ekawinataa) [inline src/main/java/com/gotocompany/firehose/sink/grpc/GrpcSinkFactory.java:60]: just a nitpick, can we put the private method below the public one?

File: src/main/java/com/gotocompany/firehose/sink/grpc/GrpcSinkFactory.java [hunk 12]
```diff
@@ -26,22 +38,64 @@
...
+        try {
+            File certFile = createCertFileFromBase64(base64Cert);
+            SslContextBuilder sslContextBuilder = SslContextBuilder.forClient().trustManager(certFile);
```
Comment (ekawinataa) [inline src/main/java/com/gotocompany/firehose/sink/grpc/GrpcSinkFactory.java:63]: Just wondering whether it is possible to pass the encoded cert directly as ByteArrayInputStream directly without writing to a temp file? Maybe something like this

 ```
byte[] decodedBytes = Base64.getDecoder().decode(base64Cert);
        ByteArrayInputStream certInputStream = new ByteArrayInputStream(decodedBytes);
        SslContextBuilder sslContextBuilder = SslContextBuilder.forClient().trustManager(certInputStream);
        return GrpcSslContexts.configure(sslContextBuilder).build();
```

**Pull request merged**